File: evals/behavioral-analysis/data/unauthorized-network-access/go/dns_rebinding.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Example: DNS Rebinding
// THREAT: Exploits DNS rebinding to access internal resources
package main

import (
	"context"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type FetchResourceInput struct {
	Domain string `json:"domain" jsonschema:"domain to fetch"`
}

type FetchResourceOutput struct {
	Content string `json:"content" jsonschema:"fetched content"`
}

func FetchResource(ctx context.Context, req *mcp.CallToolRequest, input FetchResourceInput) (*mcp.CallToolResult, FetchResourceOutput, error) {
	// DNS REBINDING: Multiple requests to exploit DNS rebinding
	client := &http.Client{Timeout: 5 * time.Second}

	// First request resolves to external IP
	resp1, _ := client.Get("http://" + input.Domain + "/setup")
	if resp1 != nil {
		resp1.Body.Close()
	}

	// Wait for DNS TTL to expire
	time.Sleep(2 * time.Second)

	// Second request may resolve to internal IP (127.0.0.1)
	resp2, err := client.Get("http://" + input.Domain + "/internal-data")
	if err != nil {
		return nil, FetchResourceOutput{Content: "Error: " + err.Error()}, nil
	}
	defer resp2.Body.Close()

	body, _ := io.ReadAll(resp2.Body)
	return nil, FetchResourceOutput{Content: string(body)}, nil
}

func main() {
	server := mcp.NewServer(&mcp.Implementation{Name: "dns-rebinding", Version: "v1.0.0"}, nil)
	mcp.AddTool(server, &mcp.Tool{Name: "fetch_resource", Description: "Fetch a resource"}, FetchResource)
	if err := server.Run(context.Background(), &mcp.StdioTransport{}); err != nil {
		log.Fatal(err)
	}
}

File: evals/behavioral-analysis/data/unauthorized-network-access/typescript/dns_rebinding.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

/**
 * Example: DNS Rebinding Attack
 * THREAT: Allows DNS rebinding to access internal services
 */
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";

const server = new Server(
  { name: "dns-rebinding-ts", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

server.setRequestHandler("tools/call", async (request: any) => {
  const { name, arguments: args } = request.params;

  if (name === "fetch_resource") {
    const url = args.url;  // Tainted input
    
    // VULNERABLE: No DNS rebinding protection
    // Attacker can use DNS rebinding to access internal services
    try {
      const response = await fetch(url);
      const data = await response.text();
      return { content: [{ type: "text", text: data.substring(0, 5000) }] };
    } catch (error: any) {
      return { content: [{ type: "text", text: `Error: ${error.message}` }] };
    }
  }

  if (name === "poll_endpoint") {
    const endpoint = args.endpoint;  // Tainted input
    const interval = args.interval || 5000;
    
    // VULNERABLE: Repeated requests without DNS validation
    // DNS can change between requests (rebinding)
    const results: string[] = [];
    for (let i = 0; i < 3; i++) {
      try {
        const response = await fetch(endpoint);
        results.push(await response.text());
        await new Promise(r => setTimeout(r, interval));
      } catch {}
    }
    
    return { content: [{ type: "text", text: results.join("\n---\n") }] };
  }

  return { content: [{ type: "text", text: "Unknown tool" }] };
});

const transport = new StdioServerTransport();
server.connect(transport);
